miniapps/autodiff/fdual.hpp

```diff
@@ -0,0 +1,554 @@
+   FDual() : pr(0), du(0) {}
+
+   /// The constructor utilized in nested definition of dual numbers.
+   /// It is used for second and higher order derivatives.
+   template<class fltyp,
+            class = typename std::enable_if<std::is_arithmetic<fltyp>::value>::type>
+   FDual(fltyp &f) : pr(f), du(0)
+   {}
+
+   template<class fltyp,
+            class = typename std::enable_if<std::is_arithmetic<fltyp>::value>::type>
+   FDual(const fltyp &f) : pr(f), du(0)
+   {}
+
+   /// Standard constructor with user supplied inpur for both parts of the dual number.
+   FDual(tbase &pr_, tbase &du_) : pr(pr_), du(du_) {}
+
+   FDual(const tbase &pr_, const tbase &du_) : pr(pr_), du(du_) {}
+
+   FDual(FDual<tbase> &nm) : pr(nm.pr), du(nm.du) {}
+
+   FDual(const FDual<tbase> &nm) : pr(nm.pr), du(nm.du) {}
+
+   /// Return the real value of the dual number.
+   tbase prim() const { return pr; }
+
+   /// Same as prim(). Return the real value of the dual number.
+   tbase real() const { return pr; }
+
+   /// Return the dual value of the dual number.
+   tbase dual() const { return du; }
+
+   /// Set the primal and the dual values.
+   void set(const tbase &pr_, const tbase &du_)
+   {
+      pr = pr_;
+      du = du_;
+   }
+
+   void prim(const tbase &pr_) { pr = pr_; }
+
+   void real(const tbase &pr_) { pr = pr_; }
+
+   void dual(const tbase &du_) { du = du_; }
+
+   void setReal(const tbase &pr_) { pr = pr_; }
+
+
+   void setDual(const tbase &du_) { du = du_; }
+
+   FDual<tbase> &operator=(tbase sc_)
+   {
+      pr = sc_;
+      du = tbase(0);
+      return *this;
+   }
+
+   FDual<tbase> &operator+=(tbase sc_)
+   {
+      pr = pr + sc_;
+      return *this;
+   }
+
+   FDual<tbase> &operator-=(tbase sc_)
+   {
+      pr = pr - sc_;
+      return *this;
+   }
+
+   FDual<tbase> &operator*=(tbase sc_)
+   {
+      pr = pr * sc_;
+      du = du * sc_;
+      return *this;
+   }
+
+   FDual<tbase> &operator/=(tbase sc_)
+   {
+      pr = pr / sc_;
+      du = du / sc_;
+      return *this;
+   }
+
+   FDual<tbase> &operator=(const FDual<tbase> &f)
+   {
+      pr = f.real();
+      du = f.dual();
+      return *this;
+   }
+
+   FDual<tbase> &operator+=(const FDual<tbase> &f)
+   {
+      pr += f.real();
+      du += f.dual();
+      return *this;
+   }
+
+   FDual<tbase> &operator-=(const FDual<tbase> &f)
+   {
+      pr -= f.real();
+      du -= f.dual();
+      return *this;
+   }
+
+   FDual<tbase> &operator*=(const FDual<tbase> &f)
+   {
+      du = du * f.real();
+      du = du + pr * f.dual();
+      pr = pr * f.real();
+      return *this;
+   }
+
+   FDual<tbase> &operator/=(const FDual<tbase> &f_)
+   {
+      pr = pr / f_.real();
+      du = du - pr * f_.dual();
+      du = du / f_.real();
+      return *this;
+   }
+};
+
+// non-member functions
+// boolean operations
+template<typename tbase>
+inline bool operator==(const FDual<tbase> &a1, const FDual<tbase> &a2)
+{
+   return a1.real() == a2.real();
+}
+
+template<typename tbase>
```

Comment:
>     1. Does the functor class need to be a template parameter or can you pass it by reference to the `AutoDiff` classes? It would avoid the need to construct a functor object at every call.

The functor class has to be templated in order to minimize code replication and to maximize performance (see the explanation in the other comment about templating the functors).  A solution where users can pass the function to the constructor is also implemented for lambda functions. Both approaches are available depending on the needs and the preferences of the users.

Reply:
Documented